src/aggregator/server/m3msg/server.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3msg
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/m3db/m3/src/aggregator/aggregator"
+	"github.com/m3db/m3/src/metrics/encoding"
+	"github.com/m3db/m3/src/metrics/encoding/protobuf"
+	"github.com/m3db/m3/src/metrics/generated/proto/metricpb"
+	"github.com/m3db/m3/src/msg/consumer"
+	"github.com/m3db/m3/src/x/instrument"
+	xserver "github.com/m3db/m3/src/x/server"
+
+	"go.uber.org/zap"
+)
+
+type server struct {
+	aggregator aggregator.Aggregator
+	logger     *zap.Logger
+}
+
+// NewServer creates a new M3Msg server.
+func NewServer(
+	address string,
+	opts Options,
+	aggregator aggregator.Aggregator,
+	instrumentOpts instrument.Options,
+) (xserver.Server, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	s := &server{
+		aggregator: aggregator,
+		logger:     instrumentOpts.Logger(),
+	}
+	handler := consumer.NewConsumerHandler(s.Consume, opts.ConsumerOptions())
+	return xserver.NewServer(address, handler, opts.ServerOptions()), nil
+}
+
+func (s *server) Consume(c consumer.Consumer) {
+	var (
+		pb     = &metricpb.MetricWithMetadatas{}
+		union  = &encoding.UnaggregatedMessageUnion{}
+		msgErr error
+		msg    consumer.Message
+	)
+	for {
+		msg, msgErr = c.Message()
+		if msgErr != nil {
+			break
+		}
+
+		err := s.handleMessage(pb, union, msg)
+		if err != nil {
+			s.logger.Error("could not process message", zap.Error(err))
+		}
+	}
+	if msgErr != nil && msgErr != io.EOF {
+		s.logger.Error("could not read message", zap.Error(msgErr))
+	}
+	c.Close()
+}
+
+func (s *server) handleMessage(
+	pb *metricpb.MetricWithMetadatas,
+	union *encoding.UnaggregatedMessageUnion,
+	msg consumer.Message,
+) error {
+	defer msg.Ack()
```

Comment:
Are all of the below error cases non-retryable? If we unconditionally ack we'll never retry from the coordinator client to the aggregator right?

Reply:
I believe we probably should avoid retries on this, if aggregator gets overloaded the amount of buffering/retries will probably quickly overwhelm the coordinators.